src/mrack/providers/provider.py

```diff
@@ -101,25 +101,39 @@ async def provision_hosts(self, hosts):
         logger.info("All hosts reached provisioning final state (ACTIVE or ERROR)")
         logger.info(f"Provisioning duration: {provi_duration}")
 
-        errors = self.parse_errors(server_results)
+        hosts = [self.to_host(srv) for srv in server_results]
+        errors = self.parse_errors(hosts)
+
         if errors:
             logger.info("Some host did not start properly")
-            for err in errors:
-                self.print_basic_info(err)
+            for host_err in errors:
+                self.print_err_info(host_err)
+
             logger.info("Given the error, will delete all hosts")
-            await self.delete_hosts(server_results)
+            await self.delete_hosts(hosts)
             raise ProvisioningError(errors)
 
-        hosts = [self.to_host(srv) for srv in server_results]
+        logger.info("Printing provisioned hosts")
         for host in hosts:
             logger.info(host)
+
         return hosts
 
-    def parse_errors(self, server_results):
+    def parse_errors(self, hosts):
         """Parse provisioning errors from provider result."""
-        raise NotImplementedError()
+        errors = []
+        for host in hosts:
+            logger.info(f"Host: {host.id}\tStatus: {host.status}")
+            if self.STATUS_MAP.get(host.status, STATUS_OTHER) == STATUS_ERROR:
+                errors.append(host.error)
+
+        return errors
+
+    def print_err_info(self, err):
```

Comment:
In the end, do we really need this method?

Reply:
In the beginning. parse_errors method was not about extracticting error messages. But it was about getting the provisioning results where host were not provisioned correctly so that we can decide what to do with it. The implemented default method was to print the error information (thus the print_err_info method) and delete all provisioned hosts. But a different provisioning strategy can delete just the hosts in error and re-try their provisioning.

Reply:
I have removed this method for now.